### ISSUE TITLE
[Bug Kiwix 3.6.0]Infinite loading icon in library tab 

### ISSUE THREAD

**Issue by Mitia99** [kiwix/kiwix-android#2445]
**Describe the bug**
Just an UI bug:Infinite loading icon in library tab
No issue with the library,zims are loaded.

**Expected behavior**
No more loading icon when zim are loaded.

**Steps to reproduce the behavior:**
1. Download zim content
2. Open library tab

**Screenshots**
![AddText_10-16-06.04.14.jpg](https://user-images.githubusercontent.com/56766778/96275742-c2c66080-0fda-11eb-9cb7-231e80801b3d.jpg)

**Environment**
- Version of Kiwix Android : 3.4.1
- Device : Huawei P10
- OS version : Android 9

**Comment by macgills** [comment]
Thank you for the report, will investigate and hopefully have a fix in not too long

**Comment by s-ayush2903** [comment]
@Mitia99 I'm sorry but I can't reproduce the issue, I tried ~5 times the procedure mentioned in the description but couldn't find it. Can you please answer one question: Does it always occur(via following the steps given in the issue desc.) or you just caught it once?

**Comment by Mitia99** [comment]
@s-ayush2903 
I tried force stop and clear cache but same issue.

[]()https://drive.google.com/file/d/1XKecR3z52jRgwCntXWLmA9xBakZZHoVB/view?usp=drivesdk

**Comment by Mitia99** [comment]
@s-ayush2903 
It always occurs.

Additionnal information:
I did an update from version 3.3.1(i.e I downloaded the zim contents with 3.3.1).

**Comment by s-ayush2903** [comment]
No worries, I'll take a look

**Comment by Mitia99** [comment]
@s-ayush2903 
I tried to clear app data(which removed all zim contents) but same issue.So I don't think it that the issue is linked with the update![Screenshot_20201021_061917.jpg](https://user-images.githubusercontent.com/56766778/96669251-880b5200-1365-11eb-8945-397899c4e95e.jpg)

**Comment by macgills** [comment]
Ahhh this is library, not Download, d'oh. So what this is doing is scanning your filesystem looking for zims. I am going to assume you are using an sd card? There is a possibility this is blocked by our file system detection, in Download do you see large entries ie >4GB, greyed out?

**Comment by Mitia99** [comment]
@macgills I'm using a 64go SD Card.Storage is set on internal![Screenshot_20201021_114225_org.kiwix.kiwixmobile.jpg](https://user-images.githubusercontent.com/56766778/96695656-91f77a00-1392-11eb-808c-5b15b518b464.jpg)

**Comment by macgills** [comment]
hmmmm so it is potentially an error in StorageObserver. Have you left the phone on this screen for an unreasonably long time? I wonder if it ever completes

**Comment by gouri-panda** [comment]
I'm encountering this problem too.

**Comment by macgills** [comment]
Would you like for this to be assigned to you then @gouri-panda ? The file search completes near instantaneously on my device

**Comment by s-ayush2903** [comment]
+1 To second line in the trailing comment, I still can't reproduce the issue

**Comment by gouri-panda** [comment]
> Would you like for this to be assigned to you then @gouri-panda ? The file search completes near instantaneously on my device

yes @macgills

**Comment by gouri-panda** [comment]
@Mitia99 I'm facing this problem but in a different way. On my phone, it loads for approximately  3-4 minutes. Do you have this situation too?

**Comment by Mitia99** [comment]
@gouri-panda Yes same here,I figured out that it is taking ~ 2minutes but when I reduce kiwix app and re-open it,it is re-scanning.

**Comment by gouri-panda** [comment]
That's because it is downloading the content even if you switched the tab, that's why it is taking so much time.

**Comment by Mitia99** [comment]
@gouri-panda Oh,I see..I hope that you could improve the scan time.

**Comment by gouri-panda** [comment]
@Mitia99 it's taking the usual time ,but if you have slow processor than you can feel the delay like I have. My one phone is taking 2 to 3 minutes and the other phone is taking hardly 10 sec.

**Comment by Mitia99** [comment]
@gouri-panda I don't think that it is a processor issue.I am using a kirin 960(8 core cpu) which is pretty fast even in huge apps.

**Comment by gouri-panda** [comment]
@Mitia99 I looked over your storage.My hypothesis is that you have a lot of content in both your internal and external memory card so thats why it scans the  whole thing after you dowload  zim files.

**Comment by Mitia99** [comment]
@gouri-panda Yes,you're right.That's why making the engine to only scan specific folders/files types will be great.For users with large files in 128go/256go memory/SD Card it will take a very long time.

**Comment by gouri-panda** [comment]
@Mitia99 It will break us our certain feature like sharing our Zim files and download the Zim files over the internet. It won't add these files into library.

**Comment by macgills** [comment]
https://github.com/kiwix/kiwix-android/issues/2405 there is a solution outlined here

**Comment by Mitia99** [comment]
@macgills Thank you for the reference.Personnaly,I don't really want a manual folder selection but only an improved auto-scan(modify a bit the way that the engine scans).

**Comment by macgills** [comment]
It is quite difficult to "intelligently" scan as any number of apps eg torrent clients, browsers, dedicated downloaders, could be downloading files and be putting them anywhere on disk as could a user transferring manually. Our only guarantee is to look everywhere and really it shouldn't be that slow but here we are.
We have 2 options as I see it and we could do both of them.

1. coroutines rewrite to try and improve performance
1. a switch to deactivate full scanning and only look at Kiwix app folders

long term we could look at user configurable search locations but sooner or later we are going to have to start supporting the SAF which is going to change file access quite a lot and there is a lot of work that will go into UI/UX of that feature (unless we just give users a textbox and let them add any path they want but I imagine that being confusing and unfriendly) and that might get trashed, wasting dev time.

**Comment by gouri-panda** [comment]
@macgills Option 1 sounds good. Should I start working on this ticket?

**Comment by macgills** [comment]
make a sub ticket for a coroutines rewrite would be best I think

**Comment by kelson42** [comment]
@gouri-panda Not sure about the status of this bug, but switching to OPDS should solve the problem.

**Comment by Mitia99** [comment]
Bug details update
•Device: Huawei P30 (no sd card)
•Android 10
•Kiwix: 3.6.0

**Comment by kelson42** [comment]
@Mitia99 Thx for the update

**Comment by Mitia99** [comment]
@kelson42 No problem. Btw, I noticed that rather than an infinite loading it is actually a slow library load.(My kiwix library size is about 700mo) Maybe it has something to do with my heavy memory usage (90% of 128go).

**Comment by kelson42** [comment]
@gouri-panda Domyou still have this problem with 3.6.0+?

**Comment by kelson42** [comment]
@gouri-panda Domyou still have this problem with 3.6.0+?

**Comment by gouri-panda** [comment]
@kelson42 We can close this issue now.

**Comment by kelson42** [comment]
@gouri-panda Why?

**Comment by gouri-panda** [comment]
I haven't encountered this issue recently. I have marked that, but I didn't know that this ticket was still open  We may be fixed this issue in the storage issues.

**Comment by Mitia99** [comment]
@gouri-panda Can you re-open this?

It's back in **3.8.0**

**Comment by MohitMaliFtechiz** [comment]
@kelson42 This issue seems duplicate of https://github.com/kiwix/kiwix-android/issues/3646. The user has the same issue as mentioned in https://github.com/kiwix/kiwix-android/issues/3646.

**Comment by MohitMaliFtechiz** [comment]
> @gouri-panda Can you re-open this?
It's back in 3.8.0
Device : Huawei P30 (no sd card)
OS version : Android 10
Thanks for investigating,

@kelson42 in the last update user confirmed that there is no SD card present in his phone. So here scanning is only performed inside the phones's storage, and he still facing the slow loading issue. Therefore, the issue is not with the SD card, it is with large numbers of files available in the storage because we scan the whole storage to find the ZIM files from every possible location of the storage.

> @gouri-panda Yes same here, I figured out that it is taking ~ 2minutes.
@kelson42 No problem. Btw, I noticed that rather than an infinite loading it is actually a slow library load.(My kiwix library size is about 700mo) Maybe it has something to do with my heavy memory usage (90% of 128go).


The user confirms that there is a slow loading in the Library screen and that 90% of his storage is full(So there is a high probability that many files are available in his storage). I had tried to put 20K dummy files(with the size of 10KB) in my system https://github.com/kiwix/kiwix-android/issues/3650#issuecomment-1875368514 to slow down the scanning while fixing the https://github.com/kiwix/kiwix-android/issues/3650, and at that time it took almost 19 seconds to load the library. Here user's device takes 2 minutes to scan and show the ZIM files in the Library screen, and we are facing the same issue in https://github.com/kiwix/kiwix-android/issues/3646.

**Comment by MohitMaliFtechiz** [comment]
> Not really. Kiwix show the ZIMs instantly and I can immediatly browse them after opening the app.

@Mitia99 you were facing a slow-scanning issue as you mentioned earlier. Happy to know now you are not facing that issue. 

> There's no major performance issues. It's more of a bad UI experience.

For this, we have improved our UI, and enhanced the scanning process in https://github.com/kiwix/kiwix-android/pull/3674.
